test/system/src/test/java/io/pravega/test/system/ReaderCheckpointTest.java

```diff
@@ -0,0 +1,242 @@
+    private final ScheduledExecutorService executor = ExecutorServiceHelpers.newScheduledThreadPool(1, "checkPointExecutor");
+    private final StreamConfiguration streamConfig = StreamConfiguration.builder()
+                                                                        .scalingPolicy(ScalingPolicy.fixed(1)).build();
+
+    private URI controllerURI;
+
+    @Environment
+    public static void initialize() throws Exception {
+
+        //1. check if zk is running, if not start it
+        Service zkService = new ZookeeperService("zookeeper");
+        if (!zkService.isRunning()) {
+            zkService.start(true);
+        }
+
+        List<URI> zkUris = zkService.getServiceDetails();
+        log.debug("Zookeeper service details: {}", zkUris);
+
+        //get the zk ip details and pass it to bk, host, controller
+        //2, check if bk is running, otherwise start, get the zk ip
+        Service bkService = new BookkeeperService("bookkeeper", zkUris.get(0));
+        if (!bkService.isRunning()) {
+            bkService.start(true);
+        }
+        log.debug("Bookkeeper service details: {}", bkService.getServiceDetails());
+
+        //3. start controller
+        Service conService = new PravegaControllerService("controller", zkUris.get(0));
+        if (!conService.isRunning()) {
+            conService.start(true);
+        }
+        List<URI> conUris = conService.getServiceDetails();
+        log.debug("Pravega Controller service details: {}", conUris);
+
+        //4.start host
+        Service segService = new PravegaSegmentStoreService("segmentstore", zkUris.get(0), conUris.get(0));
+        if (!segService.isRunning()) {
+            segService.start(true);
+        }
+        log.debug("Pravega segment store details: {}", segService.getServiceDetails());
+    }
+
+    @Before
+    public void setup() throws URISyntaxException {
+        controllerURI = fetchControllerURI();
+        StreamManager streamManager = StreamManager.create(controllerURI);
+        assertTrue("Creating Scope", streamManager.createScope(SCOPE));
+        assertTrue("Creating stream", streamManager.createStream(SCOPE, STREAM, streamConfig));
+    }
+
+    @Test
+    public void readerCheckpointTest() throws Exception {
+
+        @Cleanup
+        ReaderGroupManager readerGroupManager = ReaderGroupManager.withScope(SCOPE, controllerURI);
+        ReaderGroup readerGroup = readerGroupManager.createReaderGroup(READER_GROUP_NAME, readerGroupConfig,
+                Collections.singleton(STREAM));
+
+        int startInclusive = 1;
+        int endExclusive = 100;
+        log.info("Write events with range [{},{})", startInclusive, endExclusive);
+        writeEvents(IntStream.range(startInclusive, endExclusive).boxed().collect(Collectors.toList()));
+        readEventsAndVerify(startInclusive, endExclusive);
+
+        //initiate checkpoint100
+        Checkpoint checkPoint100 = readerGroup.initiateCheckpoint("batch100", executor).get();
```

Comment:
Is this actually going to work? With the current code, we are reading and closing the reader, but not declaring the reader offline. The reader must be able to deliver the checkpoint event, otherwise it shouldn't be able to make progress. In fact, we should also test that the checkpoint events are delivered.

Reply:
reader.close() will automatically invoke ReaderGroup#readerOffline(String, Position).

Reply:
Done. Updated the code to verify checkpoint events.